### PR TITLE
make PID watching a bit less expensive

### PR DESCRIPTION
**Pull request by poettering**
A proposed fix for #9138. Would be good if @davide125 and colleagues could give this a whirl to check that this helps a bit.

### CODE REVIEW

**Comment by davide125** [comment]
We had this running for about a day on the affected host in #9138, and so far it seems to be helping: the box is behaving normally and we're not seeing the deluge of `kill(_, 0)` in strace anymore. I'll let this sit over the weekend and report back early next week.

File: src/core/unit.c [hunk 1]
```diff
@@ -2622,6 +2627,76 @@ void unit_tidy_watch_pids(Unit *u, pid_t except1, pid_t except2) {
...
+                return r;
+        if (r > 0) /* On unified we can use proper notifications */
+                return 0;
+
+        /* Enqueues a low-priority job that will clean up dead PIDs from our list of PIDs to watch and subscribe to new
+         * PIDs that might have appeared. We do this in a delayed job because the work might be quite slow, as it
+         * involves issuing kill(pid, 0) on all processes we watch. */
+
+        if (!u->rewatch_pids_event_source) {
+                sd_event_source *s = NULL;
```
Comment (keszybz) [inline src/core/unit.c]: This needs `_cleanup_`.

**Comment by keszybz** [comment]
Hmm, `bionic-amd64` failed with a timeout in "upstream". Those tests are pretty flaky, but I haven't seen this failure yet. Let's see what happens when it runs again.

**Comment by davide125** [comment]
Update: our test host is still doing fine with this PR applied

File: src/core/unit.c (hunk 1 above)
Comment (poettering) [reply src/core/unit.c]: ouch! indeed!

**Comment by poettering** [comment]
@davide125 thanks for reporting back!

@keszybz thanks for the review!

force pushed a new version now. Only change: the _cleanup_ decorator is now added, fixing the memory leak i introduced, as pointed out by @keszybz. PTAL!

**Pull request merged**